### ISSUE TITLE
Lock nodes offset

### ISSUE THREAD

**Issue by Stridemann** [ReClassNET/ReClass.NET#48]
Hi,
Thanks again for a program.

I just wanted to ask if it's possible to add LOCK node feature. When you lock node it's offset can't be broken after deleting/inserting bytes before/under it.
Sometimes we don't need to define types in whole structure, just some nodes at start, middle and end of structure.

Also this feature gives us other new awesome feature - we can delete some byte range we don't need (like "hide" in old reclass).

I know, this can be not easy task, just wanted to ask if this is possible.
Thanks you.

**Comment by Stridemann** [comment]
Also adding options for inserting 1 and 2 bytes will be awesome too.

**Comment by KittoniuM** [comment]
ye, locking something would be amazing, like the checkbox when u add a value in cheat engine, u press it, and it forces the value to it every 50 milisecond or whatever, pls knacker!!

**Comment by KN4CK3R** [comment]
> ye, locking something would be amazing, like the checkbox when u add a value in cheat engine, u press it, and it forces the value to it every 50 milisecond
That's not what he is talking about.

**Comment by KN4CK3R** [comment]
@Stridemann Currently that's not possible. There is a rewrite of the internal class representation planned. The new system may have support for that.

**Comment by Stridemann** [comment]
@KN4CK3R Thanks a lot for your work on this project. That will be really great.

I got one stable bug that throws crash error (can continue working, no data lost). Don't want to create a new issue due to low priority: When you double click on node to edit name (enable text input field) and then right click on any other place or node you will got error.

**Comment by KN4CK3R** [comment]
> I got one stable bug that throws crash error (can continue working, no data lost). Don't want to create a new issue due to low priority: When you double click on node to edit name (enable text input field) and then right click on any other place or node you will got error.

But a crash is a real issue so should have created another one.

**Comment by KN4CK3R** [comment]
The original request isn't implemented...

**Comment by VisualCH** [comment]
Still no fix for this?

**Comment by Timboy67678** [comment]
Nothing to fix, this is a feature request.

**Comment by LoneDev6** [comment]
Hello!
I think this feature request is a must have to avoid doing using a top-down approach when reversing a structure.
Let us know what you think @KN4CK3R

**Comment by oculus472** [comment]
Does anyone know of a good way to handle this? At the moment redefining fields is a nightmare.

**Comment by LoneDev6** [comment]
No idea sincerely, all the tools I tried have the same problem.
I always have to redefine all my structures from the ground up

**Comment by ross-weir** [comment]
@KN4CK3R Is this still something you'd be interested having added?

If so I'd like to have a go at implementing this. Do you have any preferences/suggestions on how to go about it?

From a quick look I was thinking something like:
- Before adding nodes in `InsertBytes` save all defined node offsets - defined nodes being those with a non empty `Name` value.
- After adding nodes restore saved nodes to original offsets

**Comment by KN4CK3R** [comment]
You have to keep in mind that `BaseNode.Offset` is not intended to be frozen. Every action which changes the class layout recalculates the offset for all nodes.

**Comment by LoneDev6** [comment]
@KN4CK3R  So basically this means that the program is structured in a way that this feature is very hard to implement without a lot of modifications?

**Comment by KN4CK3R** [comment]
Yes, you would have to insert `HexNodes` infront to increase a nodes offset or remove them. Removing will be a problem if there are not unused hex nodes to remove.
Something like this is already there:
https://github.com/ReClassNET/ReClass.NET/blob/0ee8a4cd6a00e2664f2ef3250a81089c32d69392/ReClass.NET/Nodes/BaseContainerNode.cs#L191-L204
Commented out 5 years ago because I decided it's not worth it.
